### ISSUE TITLE
Has the updater ever worked since 2.7?

### ISSUE THREAD

**Issue by Killercrusher232** [Gang-Garrison-2/Gang-Garrison-2#157]
````
ERROR in
action number 1
of Create Event
for object Updater:

Error in code at line 51:
           progressBar = floor((progress/size) * 20);
                                         ^
at position 40: Division by 0.
````

Last time I tried to update it just did nothing, now it actually errors too.

**Comment by wareya** [comment]
top kek

**Comment by Medo42** [comment]
This happened because `http_response_body_size(downloadHandle)` returned 0. I don't know under which conditions that can happen, but in any case we should make it so that it won't cause a crash. I suggest we change the condition in the code that updates the progress bar from `if (size != -1)` to `if (size > 0)`.

**Comment by TazeTSchnitzel** [comment]
This sounds like a Faucet HTTP issue. Hmm...

**Comment by TazeTSchnitzel** [comment]
OK, `http_response_body_size` just returns `client.responseBodySize` in Faucet HTTP. That's set in only two places. The first is [when we see a "content-length" header"](https://github.com/TazeTSchnitzel/Faucet-HTTP-Extension/blob/9b5c9942d540e60a3b64509ebe5d2d4151a2afc6/faucet-http.gml#L746) and the second is [when we decode a chunked-transfer body](https://github.com/TazeTSchnitzel/Faucet-HTTP-Extension/blob/9b5c9942d540e60a3b64509ebe5d2d4151a2afc6/faucet-http.gml#L1003). In neither situation should it be zero. If the `Content-Length` parsed to zero, we're downloading a zero-length file. It can't be `real()` mangling data, that function just throws an error if the string is suspect. Similarly, I don't think a chunked encoding could result in zero, unless it's a zero-length file.

Perhaps the client is somehow fetching the wrong file?

**Comment by TazeTSchnitzel** [comment]
Oh, I think I figured out why this is happening. `UPDATE_SOURCE` is http://www.ganggarrison.com/gg2update.zip which is a redirect. HTTP redirect responses don't need bodies, and indeed, that has a zero-length body. Faucet HTTP handles redirects by restarting the request once it finds out, so initially, GG2 is told the body is zero-length, because it's seeing the redirect body and Faucet HTTP hasn't followed the redirect yet.

So, the solution is, as MedO mentions, to check for `> 0` instead of `!= -1`.

**Comment by TazeTSchnitzel** [comment]
Oh dear. Fixing the div0 thing results in Faucet HTTP crashing somewhere else, due to a type error. Oh GML, how I love thee.

**Comment by TazeTSchnitzel** [comment]
Error was caused by me changing the signature of a function (`__http_prepare_request`), but failing to update all call sites. Obviously it's also my fault for not grepping, but this is another example of why lack of static type checking is a pain.

This is fixed in [Faucet HTTP v1.1.1](https://github.com/TazeTSchnitzel/Faucet-HTTP-Extension/releases/tag/V1.1.1), and [here's a pull request to update GG2's version](https://github.com/Medo42/Gang-Garrison-2/pull/281).

**Comment by TazeTSchnitzel** [comment]
tl;dr: GML sucks but it's mostly my fault

**Comment by Medo42** [comment]
Mergerino

**Comment by TazeTSchnitzel** [comment]
You only merged one half of the fix

**Comment by Medo42** [comment]
Good to know, but I'm pretty sure I'll change that in a minute :)

**Comment by TazeTSchnitzel** [comment]
Sure enough, you did!
